src/k8s/pkg/snap/snap.go

```diff
@@ -0,0 +1,102 @@
+package snap
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/canonical/k8s/pkg/utils"
+)
+
+// snap implements the Snap interface.
+type snap struct {
+	snapDir       string
+	snapDataDir   string
+	snapCommonDir string
+}
+
+// NewSnap creates a new interface with the K8s snap.
+// NewSnap accepts the $SNAP, $SNAP_DATA and $SNAP_COMMON, directories
+func NewSnap(snapDir, snapDataDir, snapCommonDir string, options ...func(s *snap)) Snap {
+	s := &snap{
+		snapDir:       snapDir,
+		snapDataDir:   snapDataDir,
+		snapCommonDir: snapCommonDir,
+	}
+
+	for _, opt := range options {
+		opt(s)
+	}
+	return s
+}
+
+var snapInstance Snap
+var lock = &sync.Mutex{}
+
+// Instance returns an instance of the snap.
+//
+// FIXME: We don't have a single entrypoint like in the cluster-agent.
+// However, microcluster plans to add an OnRequest hook in the future where
+// we could inject that snap struct in a single place at app start.
+//
+// Singletons have a lot of downsides but I think
+// this fits quite well here. The snap struct doesn't really
+// handle any state and with the given structure of the k8s snap
+// it is not easy to create and inject one snap instance in the
+// application root.
+func Instance() Snap {
+	if snapInstance == nil {
+		lock.Lock()
+		defer lock.Unlock()
+		if snapInstance == nil {
+			snapInstance = NewSnap(
+				os.Getenv("SNAP"),
+				os.Getenv("SNAP_DATA"),
+				os.Getenv("SNAP_COMMON"),
+			)
+		}
+	}
+
+	return snapInstance
```

Comment:
Is this semantic a go norm to double check the `inst` value?  I don't have much experience here in go so i'm just asking


I did find another pattern using `sync.Once`

```go
var once sync.Once

type single struct {
}

var singleInstance *single

func getInstance() *single {
    if singleInstance == nil {
        once.Do(
            func() {
                fmt.Println("Creating single instance now.")
                singleInstance = &single{}
            })
    } else {
        fmt.Println("Single instance already created.")
    }

    return singleInstance
}
```

Reply:
make the instance a pointer instead and use `once.Do` for the instance creation

```go
var snapInstance *Snap
```


```suggestion
    once.Do(func() {
    	snapInstance = &NewSnap(
		    os.Getenv("SNAP"),
			os.Getenv("SNAP_DATA"),
			os.Getenv("SNAP_COMMON"),
		)
    }
	return snapInstance
```

Reply:
`sync.Once` is indeed a way to do this, but see comment above, we don't really need this.

Reply:
The double `nil` check is because:
```
There is another nil-check after the lock is acquired. This is to ensure that if more than one goroutine bypasses the first check, only one goroutine can create the singleton instance. Otherwise, all goroutines will create their own instances of the singleton struct.
```

(shamelessly copied from here: https://refactoring.guru/design-patterns/singleton/go/example)

Your suggestion works as well (thanks for that!) but this is kind of a temporary solution as Angelos already wrote, so I'll keep my variant. Also note that all this locking is required to avoid problems with a massive amount of go routines, which we do not have.